### ISSUE TITLE
Can this be used commercially?

### ISSUE THREAD

**Issue by juangea** [microsoft/TRELLIS#41]
Can this be used commercially?
That's the question, thanks!

**Comment by bitplane** [comment]
This is likely a tough question. The LICENSE says you can, but it does not have patent waivers. So it'll ultimately depend on whether Microsoft filed any patents and whether they cover your product's use cases.

**Comment by juangea** [comment]
Maybe @sicxu , @YuDeng or @JeffreyXiang can confirm this can be used commercially.

But thanks for the input @bitplane :)

**Comment by YuDeng** [comment]
Hi @juangea, our pretrained models and most of our code are available under the MIT license, which permits commercial use with the only requirements being the preservation of copyright and license notices. However, the components related to rendering—such as rendering 3D Gaussians, Radiance fields, and extracting meshes and GLBs—are not available for commercial use. You will need to rewrite these parts if you intend to use them commercially. For more details, please refer to our license information at [this link](https://github.com/Microsoft/TRELLIS?tab=readme-ov-file#%EF%B8%8F-license).

**Comment by juangea** [comment]
Thanks @YuDeng 

It seems FlexiCubes is now part of Kaolin (The core functions of FlexiCubes are now in [Kaolin](https://github.com/NVIDIAGameWorks/kaolin/) starting from v0.15.0) so maybe that's an alternative, (it's under Apache 2.0) 

I'm not sure what could be a good alternative to Differential Gaussian Rasterization.

In any case, since I'm not a low level programmer, this may be out of my capabilities, but I want to explore the possibilities, those libraries are used as a postproduction of the model output? 

In what phases or what order are these libraries used?

I mean, if we use the model, we could investigate a different way of rendering the gaussian splatting or the glb extraction, but I'm not entirely sure what the output if the model is.

Cna you clarify this for me please? 

No need for super long explanation, I don't want to steal too much time from you, just the basics to understand what are talking about here.

Thanks a lot!

**Comment by YuDeng** [comment]
The direct outputs of our models are: Gaussian primitives, local radiance fields within active voxels, or local signed distance values and parameters for Flexicubes to extract mesh surfaces. Up to these parts, everything follows the MIT license. After that, for rendering Gaussians, radiance fields, and extracting meshes, they are under different licenses and should be replaced.

**Comment by jclarkk** [comment]
@juangea 

Even in kaolin it's stated that this part is non commercial:
![Image](https://github.com/user-attachments/assets/f8708cb7-5ff3-40fa-912d-a2fc013e5d30)

**Comment by jclarkk** [comment]
@juangea 
I do have some good news though, I managed to implement the renderer using [gsplat](https://github.com/nerfstudio-project/gsplat) which has a permissive license.

@YuDeng 
Would you like a PR on this?

**Comment by YuDeng** [comment]
Hi @jclarkk, thanks for that. You can go ahead to make the PR or share the link of your fork here.

For now, I think it's better to keep this repo as is as a verified research project. We'll look into gsplat and evaluate its performance later (for both inference and training).

**Comment by DiamondGlassDrill** [comment]
@jcclark would love to get to know more about your fork and gsplat. Thanks in advance

**Comment by tommi-pirttiniemi** [comment]
What about license of the output 3d models?

**Comment by YuDeng** [comment]
Hi @tommi-pirttiniemi, when exporting GLB files, the process involves using Flexicubes to extract meshes and a non-commercial [Gaussian renderer](https://github.com/graphdeco-inria/diff-gaussian-rasterization) for baking the albedos. It also involves using [pymeshfix](https://github.com/pyvista/pymeshfix) for mesh post-processing, which is under a [GPL-3.0 license](https://github.com/pyvista/pymeshfix#GPL-3.0-1-ov-file) and allows for commercial use.

As suggested by @jclarkk in https://github.com/microsoft/TRELLIS/pull/73, [gsplat](https://github.com/nerfstudio-project/gsplat) can be an alternative to the Gaussian renderer; however, replacing the Flexicubes part requires further effort.

**Comment by jclarkk** [comment]
@juangea @YuDeng 
I've also implemented Marching Cubes as a replacement for FlexiCubes. You can view the PR here:
https://github.com/microsoft/TRELLIS/pull/89

**Comment by juangea** [comment]
Awesome @jclarkk!!

I presume the PR won’t be accepted here being an official repo, is everything usable from your repo?

**Comment by jclarkk** [comment]
@juangea 
Yes, you can simply use my [main](https://github.com/jclarkk/TRELLIS) branch as it contains all the changes I've mentioned here.

**Comment by 0lento** [comment]
As far as I know, we can now get as far as decoding gaussian and and raw mesh with permissive licenses, thanks to jclarkk.
Decimation is already MIT (regular pyvista) and also Trimesh (MIT) can do that as well.

I've been testing Trellis by fully removing the problematic libraries and found that there is still nvdiffrast dependency left for Utils3D and random direct nvdiffrast calls that get used at:
- MeshRenderer (when rendering mesh normals video preview for example, which could just be skipped)
- Fill Holes/invisible removal (I'd debate this isn't really mandatory either, as it seems to also make holes but it sure gets rid of extra polys, there's also potential for other, faster filtering options here?)
- Texture baking (both Fast and Optimized modes need nvdiffrast atm)

Also if you need radiance fields, it seems to have diffoctreerast dependency which brings another limiting license, but afaik we don't need octrees or radiance fields to get the meshes or textures out?

In the big picture, I'd say that texturing phase is most crucial here to swap. There are a lot of MIT/BSD licensed options for differential renderers, but I don't really know which ones would be best suited for this. This would need some testing because speed and VRAM requirements play a notable role at this stage, at least with current baking strategy. Would SoftRas / DEODR / PyTorch3D etc work here?

There's also potential here for different texture projection strategies than what is used now, but just porting different renderer for the current texture bake would probably be simplest to implement. Did TRELLIS researchers experiment with other ways to accomplish the texture projection?

Also curious if it would be possible to just extract the raw mesh with colors applied to vertex colors for quick mesh visualizations?

**Comment by jclarkk** [comment]
@0lento 
Thanks for mapping out all the tricky parts, it's really helpful.

Normally, projects utilize PyTorch3D for the texturing process, so I might try it there and attempt to replace the baking process.

**Comment by MooSpyker** [comment]
Is anyone working on a way to export using another MIT license software? Or are we screwed using 3D assets in video games using this tool all together? Its weird that Trellis itself is commercial but just not the exporter... how else could you use Trellis commercially then... without getting at the actual output models? Makes no sense to me.

**Comment by 0lento** [comment]
> Its weird that Trellis itself is commercial but just not the exporter...

Trellis is a research project. Commercial use isn't the aim of such projects. Trellis team has given us very generous MIT license for their part though (and it's great). Note that it's not just the exporter that originally uses non-commercial license. So does the originally used gaussian renderer, radiance fields, mesh generation and mesh rendering.

Now today, afaik you can get the actual mesh out using MIT compatible pipeline (for code side) if you use jclarkk's Marching Cubes PR. Even decimating it isn't an issue. Disclaimer: I haven't checked each involved license to that mentioned point manually, so that would still need some verification, we are still mapping this situation out with more broad strokes right now to see what all needs to change.

As mentioned on my previous post, current texture bake and some of the mesh cleanup are still using noncommercial licenses, but that part is definitely solvable one way or another. I'm fairly sure for example that you could even omit the whole gaussian step and sample the texture directly using structured latent data. Or just swap the current texture bake to use different renderer.

But the code licensing isn't really the only question mark here. To me the biggest question here is, how are these licenses actually going to affect the generated output or the weights license? 

MS gives MIT license for their part, but hasn't the public weights been trained with code libraries that have noncommercial clause in them? Also I think 3D-Future and HSSD-Models datasets don't allow commercial use. This would depend on the license MS got for them or if they are under same NC license that everyone else gets.

I'm no lawyer but I'd assume to truly make Trellis commercially viable, you'd also need to train the models using datasets that allow commercial use, for example by only using Objaverse XL (which has Apache license). I'm not sure if I've seen anywhere how much compute resources did it take to train Trellis models from scratch but I'm just going to assume it's not something that's out of reach for most here.

**Comment by MooSpyker** [comment]
So basically right now its impossible to use anything this tool outputs as a video game asset? Are they working towards that, or is this a tech demo thats basically out and done being worked on?

(I still have not been able to get this working in ComfyUI on my end after trying to the last 3 days... pretty frustrated at this point, we REALLY need an installation tutorial, none that are out right now get you even close...)

**Comment by juangea** [comment]
The model has been trained with a MIT license dataset that has been made public AFAIK so the model is MIT.

Also @jclarkk made some improvements to change the problematic components, however I’m not sure if the texture extraction is under that modification, but AFAIK for the mesh, it should be Ok

**Comment by 0lento** [comment]
> The model has been trained with a MIT license dataset that has been made public AFAIK so the model is MIT.

I can't find such information about MIT datasets on the project page anywhere, instead I find this:

"We provide TRELLIS-500K, a large-scale dataset containing 500K 3D assets curated from [Objaverse(XL)](https://objaverse.allenai.org/), [ABO](https://amazon-berkeley-objects.s3.amazonaws.com/index.html), [3D-FUTURE](https://tianchi.aliyun.com/specials/promotion/alibaba-3d-future), [HSSD](https://huggingface.co/datasets/hssd/hssd-models), and [Toys4k](https://github.com/rehg-lab/lowshot-shapebias/tree/main/toys4k), filtered based on aesthetic scores."

If you follow the links, you'll find that Objaverse XL is Apache License, ABO is CC-BY (doesn't prevent commercial use but need to credit), 3D-Future has NC license, so does HSSD and Toys4K may have assets that have CC-BY-NC-SA license (their licenses vary per asset).

**Comment by colortelevision** [comment]
ask permission later

**Comment by rawwerks** [comment]
thank you so much @jclarkk for doing all of this work to make the entire end-to-end pipeline more permissive!

**Comment by kg-git-dev** [comment]
I have expanded on the excellent work by @jclarkk and replaced nvdiffrast with pytorch3d. The generated mesh can be now used commercially.
Link here:
https://github.com/kg-git-dev/trellis-refactored/

**Comment by iernstig** [comment]
apparently, flexicubes from kaolin is now apache-2.0! https://github.com/NVIDIAGameWorks/kaolin/pull/883/files

**Comment by iernstig** [comment]
@kg-git-dev I'm not really sure it's so simple to say? Does trellis not also use RMBG (at least for the gradio demo) https://huggingface.co/briaai/RMBG-2.0 where you may buy the license if you want to use it.

**Comment by iernstig** [comment]
never mind my previous comment. It was not rmbg but rembg which is used https://github.com/danielgatis/rembg

**Comment by astradamus** [comment]
@0lento @jclarkk @kg-git-dev @juangea 


Not sure if anyone is still following this, but:

To clarify, I am not a lawyer! This is not legal advice! (lol) but from what I can gather there is no reason to believe that the license terms of nvdiffrast constrain the use of Trellis output models in commercial works.

See https://github.com/NVlabs/nvdiffrast?tab=License-1-ov-file :

>     The terms "reproduce," "reproduction," "derivative works," and
>     "distribution" have the meaning as provided under U.S. copyright law;
>     provided, however, that for the purposes of this License, derivative
>     works shall not include works that remain separable from, or merely
>     link (or bind by name) to the interfaces of, the Work.

(Again, still not a lawyer! This is still not legal advice!) An output GLB file from Trellis is quite separable from the Trellis software itself. I am pretty sure that this text is specifically carving out this exact situation.

In other words, I believe this license constrains Trellis itself from being sold. I think this would also constrain, say, a paid online service for producing models from images for end users that relies on Trellis. In that case, you would be monetizing the software itself (Trellis being a derivative work of nvdiffrast), rather than merely its outputs (which the license explicitly carves out as NOT derivative works).

I have reached out to the creator of nvdiffrast for clarification and will report back if he responds.

**Comment by astradamus** [comment]
**(I am still not a lawyer and this is still not legal advice)** 

Now consider the license for the "non-commercial gaussian renderer" that @YuDeng linked.

See https://github.com/graphdeco-inria/diff-gaussian-rasterization?tab=License-1-ov-file : 

> THE USER CANNOT USE, EXPLOIT OR DISTRIBUTE THE _SOFTWARE_ FOR COMMERCIAL PURPOSES
> WITHOUT PRIOR AND EXPLICIT CONSENT OF LICENSORS.

Those italics are NOT mine. They explicitly italicize the word "SOFTWARE". Output meshes are not the software! They do not include the software in any way.

I think this license also does not constrain the use of Trellis **_outputs_** in commercial projects. I think it only constrains the use of Trellis software ITSELF in a commercial project--again the same example as above, a paid online service that uses Trellis to produce models for end users.

**(I am still not a lawyer and this is still not legal advice)** 

Then next up, Flexicubes is now fully under Apache 2.0 so that's no longer a constraint.

What modules remain that might constrain commercial use?
